Story/google analytics

ksiegel1923 opened the pull request.

ksiegel1923 committed:
initialized ga4 setup

ksiegel1923 committed:
added google analytics events that track current page as the screen title and also triggers a page change event

ksiegel1923 committed:
moved google analytics so tracks for every page change not just if there is a form to submit

ksiegel1923 committed:
moved GA functions from module and into separate file, added extra event for future user research

ksiegel1923 committed:
added google analytics event that tracks whether the module has come into view when it is embedded in another site

ksiegel1923 committed:
moved the useOnScreen hook into a hooks folder

ksiegel1923 committed:
cleanup before PR

ksiegel1923 committed:
merged with dev before creating PR

michellexliu commented on feedback-module/src/components/Module.js:
```diff
@@ -125,6 +143,12 @@ function Module({ pagetitle, endpoint, dir }) {
+    ga && pageTitleAsScreen(ga, t(screen.title));
+
+    //triggers a new event called page_change that shares your current page and next page
+    ga && pageChange(ga, t(screen.title), t(SCREENS[nextScreen].title));
```
Do we want the translated version or the English version of the title? Should we perhaps ask design about this?

ksiegel1923 replied:
Probably just English because the only people looking at this data will be the agency webmasters

michellexliu replied:
The changeScreen function gets called whenever a button is clicked, which doesn't always correspond to whether or not there's a new screen (i.e. if there's an error in one of the input fields, changeScreen will still get called even if the module doesn't actually move on to the next screen). We should probably move this to right before the setScreen gets called (I believe I refactored this changeScreen function quite a bit, so the placement would be something we'd probably have to figure out after merging that).

michellexliu replied:
Alternatively, I think we could also just add this to the useEffect where screen is a dependency? not 100% sure on this though

ksiegel1923 committed:
altered ga info so only in English

michellexliu commented on feedback-module/src/lib/utils/googleAnalytics.js:
```diff
@@ -0,0 +1,21 @@
+export function trackFutureResearch(ga) {
```
Instead of passing in `ga` to all of these methods, could we refactor this into a hook and just have useGA4React there since they all need `ga` anyways?

ksiegel1923 replied:
done

michellexliu commented on feedback-module/src/index.js:
```diff
@@ -4,38 +4,51 @@ import "@trussworks/react-uswds/lib/index.css";
-  );
+  Div.attributes.gaid && (gaid = Div.attributes.gaid.value);
+
+  const ga4react = new GA4React(gaid);
```
Do we need to create a new GA4React object every time the app is re-rendered (i.e. could we put this line outside of the renderApp function)? Not too familiar with how GA4React works, so I'm just wondering.

ksiegel1923 replied:
Right now we are getting the gaid attribute within the render app which is why I put this here. I guess we could make the assumption that the user won't change their gaid so we wouldn't need to observe for it?

michellexliu replied:
Hmmm that's a good point— I guess we can keep it how it is just in case?

michellexliu commented on feedback-module/src/components/Module.js:
```diff
@@ -32,13 +41,19 @@ function Module({ pagetitle, endpoint, dir }) {
 
+  const ga = useGA4React();
+
+  const ref = useRef();
```
Should probably use a more descriptive name for this ref since there are other refs in this file

ksiegel1923 committed:
PR comments: changed googleAnalytics file to hook, changed variable name of ref to be more descriptive

michellexliu commented on feedback-module/src/lib/hooks/moduleOnScreen.js:
```diff
@@ -0,0 +1,19 @@
+import { useEffect, useState } from "react";
+
+export default function moduleOnScreen(ref) {
+  const [isIntersecting, setIntersecting] = useState(false);
```
Could we move `const [userViewed, setUserViewed] = useState(false);` here, move `checkVisible()` to this file as well and change the if statement to `if (!userViewed && isIntersecting)`, and then call `checkVisible()` after line 7? I think we could also combine this hook with the googleAnalytics() hook, so that we have easy access to the userViewedModule() method.

ksiegel1923 replied:
The only problem with this is wouldn't it reset userViewed to false every time. We only want to send this event once

michellexliu replied:
I don't think so— we never call setUserViewed(false) and since we're not actually changing/reloading the page or anything like that, `moduleOnScreen(ref)` only gets called once so the state should stay the same on all the screens; I don't think it's functionally any different than having the `useState(false)` for `userViewed` in Module.js.

ksiegel1923 replied:
I'll test it out and see what happens

ksiegel1923 committed:
fixed merge conflicts with dev

michellexliu commented on feedback-module/src/components/Module.js:
```diff
@@ -130,6 +114,8 @@ function Module({ pagetitle, endpoint, dir }) {
       focusFirstError();
     } else {
       updateOtherField();
+      pageTitleAsScreen;
```
Do we not need to provide any arguments here?

ksiegel1923 committed:
added arguments to pageChange and pageTitleAsScreen calls

michellexliu commented on feedback-module/src/components/Module.js:
```diff
@@ -100,6 +114,8 @@ function Module({ pagetitle, endpoint, dir }) {
     } else {
       updateOtherField();
+      pageTitleAsScreen(en(screen.title));
+      pageChange(en(screen.title), en(SCREENS[nextScreen].title));
```
Should be `en(screen.title, { page: pagetitle })` (same for other one) bc of the string interpolation in our translation config.

ksiegel1923 committed:
PR comment: move moduleOnScreen to googleAnalytics hook

ksiegel1923 committed:
only initialize ga4react with a gaid once (avoid problems with language change)

ksiegel1923 committed:
changed parameter of renderApp for language change

michellexliu committed:
moved initialize() out of renderApp so that it's only called once

michellexliu committed:
resolved merge conflicts

ksiegel1923 committed:
bugfixes:
if there is no screen title or next screen title then look for the titleInverse
added {page: pageTitle} for english translation because of string interpolation
